### PR TITLE
feat(Mode): onModeInit hook

### PR DESCRIPTION
**Pull request by igoroctaviano**
### Context

This hook allows the initialization of the mode's data before waiting for the mode's initialization (mode's use effects).

### Changes & Results

- Add new mode hook: onModeInit
- Update documentation

### Testing

- Use onModeInit hook (logging)

#### PR

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

- [x] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [x] OS: macOS 13.4 (22F66)
- [x] Node version: v16.20.2
- [x] Browser: Chrome 116.0.5845.140

[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch

### CODE REVIEW

File: platform/core/src/extensions/ExtensionManager.ts [hunk 1]
```diff
@@ -109,6 +110,41 @@ export default class ExtensionManager extends PubSubService {
     );
   }
 
+  /**
+   * Calls all the services and extension on mode init.
+   * The service onModeInit is called first then registered extensions onModeInit is called.
+   * This is supposed to setup the extension for custom initial setup.
+   */
+  public onModeInit(): void {
```
Comment (sedghi) [inline platform/core/src/extensions/ExtensionManager.ts:118]: This is getting out of hand, can we refactor it into something like

```js
public onModeChange(mode: 'enter' | 'exit' |'init')
```

Comment (igoroctaviano) [reply platform/core/src/extensions/ExtensionManager.ts:118]: I like your suggestion, it scales better. If that's what we want I'll go ahead and refactor it.

**Review by sedghi — COMMENTED** [review]
can you clearly explain in the docs, who runs first, init or Enter? why one cannot write their logic at the start of the enter if they want to run it.

File: platform/docs/docs/platform/extensions/lifecycle.md [hunk 2]
```diff
@@ -78,6 +79,28 @@ export default {
 };
 ```
 
+## onModeInit
+
+If an extension defines the `onModeInit` lifecycle hook, it is called when a
+new mode right before the mode enters, or a mode's data or datasource is switched.
```
[the remark this answers is not in the document]
Comment (igoroctaviano) [reply platform/docs/docs/platform/extensions/lifecycle.md:85]: @sedghi the init runs before the mode has initialized (waited for useEffect + conditions). Init is when the component first mounts.

[the remark this answers is not in the document]
Comment (igoroctaviano) [reply platform/docs/docs/platform/extensions/lifecycle.md:85]: @sedghi its saying "right before the mode enters"

**Comment by igoroctaviano** [comment]
> can you clearly explain in the docs, who runs first, init or Enter? why one cannot write their logic at the start of the enter if they want to run it.

Component is being created => onModeInit (first thing that runs inside the component definition)
Component is created + init is done (respecting use effects and conditions) => onModeEnter (mode renders / enters)

File: platform/docs/docs/platform/extensions/lifecycle.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sedghi) [reply platform/docs/docs/platform/extensions/lifecycle.md:85]: I don't like how for explaining when the lifeCycle runs we are discussing the react internals (useEffect etc.). What are the use cases that require you to have another hook extra to the onModeEnter, or let me rephrase it this way, why the onModeEnter is not good enough for you to add your logic?

[the remark this answers is not in the document]
Comment (igoroctaviano) [reply platform/docs/docs/platform/extensions/lifecycle.md:85]: > I don't like how for explaining when the lifeCycle runs we are discussing the react internals (useEffect etc.). What are the use cases that require you to have another hook extra to the onModeEnter, or let me rephrase it this way, why the onModeEnter is not good enough for you to add your logic?

I see! I'm cool rephrasing it, I brought react concepts to the PR discussion just as an attempt to make things clearer (technically speaking) but in the doc I tried to stick with the language used by other documented hooks. 

Here's what I tried to do...

I wanted to select a data source (conditionally or not) before the mode entered and using the enter hook was giving me a bunch of errors which at the time were caused by setting the dataSource too late while the mode was entering which I could fix by calling it before waiting for use effects (as a new hook named init):
```
    onModeInit: ({ extensionManager }) => {
      extensionManager.setActiveDataSource("random-dicomweb");
    },
```

But now I tried again using enter + setActiveDataSource and it seems to be working 😵‍💫 
I'll close the PR for now and ratest my workflow/use cases. if I get into the same dead end I'll come back and reopen the PR based on what we have discussed so far. Thx for the review, sorry for the trouble!
